### ISSUE TITLE
Are Stores other than Bookshelf supported?

### ISSUE THREAD

**Issue by dcherman** [endpoints/endpoints#120]
I'm looking to try using endpoints, but I would need to write a custom store implementation since my datastore is actually another webservice rather than a SQL database.

From the docs on the `README` it seems like this should be possible, but I see explicit references in a lot of documentation that makes explicit references to `Bookshelf`, and it's unclear how you would implement your own store and what interface you have to implement.

I understand this is in heavy development, but is what I'm doing possible, and is there any in-code documentation ( or just plain old code ) that you could point me at to get started?

**Comment by tkellen** [comment]
Yes, stores other than Bookshelf are possible. What you're describing doing is exactly what @bobholt and I have designed this architecture to support. Unfortunately, the API for stores is not yet solidly defined--it would be a bit tricky to do this without tracking this repo very closely. I worry I might be breaking your implementation regularly if you tried.

That said, you'd basically need to duplicate this API and drop it in:
https://github.com/endpoints/endpoints/blob/master/src/store-bookshelf/index.js
^ this is about to change in #121

**Comment by dcherman** [comment]
That's the price I'll have to pay for using pre-1.0 software based on a pre-1.0 spec :monkey_face:

Just means I'll need a good test suite and run it against the repo regularly if we end up deciding to continue with JSON API.

**Comment by dcherman** [comment]
Are issues the best place to ask questions, or do you guys have an IRC room that you hang out in?  I've got a *very* barebones store implemented that just returns hard-coded data for a single model, but it works!

In order to start wiring it up against the real service, I need to pass along some authentication information that was received in the initial request.  It doesn't look like you can customize the request handler either via config or subclassing in order to extract additional information before passing it off to the DataStore, but I might've missed something too

**Comment by dcherman** [comment]
So far, my solution is a bit hackish and is probably relying on way pretty private implementation details:

```js
var RequestHandler = require( 'endpoints/es5/request-handler' ),
    query = RequestHandler.prototype.query;
    RequestHandler.prototype.query = function( request ) {
        var retValue = query.apply( this, arguments );
        retValue.request = request;
        return retValue;
    };
```

By doing that, it makes the request available for me in my `read` method to forward authentication headers, but I doubt that's how this should actually work.

**Comment by tkellen** [comment]
Heya @dcherman!

Issues are good for specific problems, IRC is probably better for general discussions. I just created #endpoints on freenode.

So, the store should not have direct access to the request object. It should only receive a normalized subset of it via `RequestHandler#query`. I haven't dealt with any auth stuff yet, but the RequestHandler should be configurable per-controller, and it should have a customizable async lookup method to determine the user of the request. That user should then become a part of the output of `#query`.

I haven't implemented any of this yet, but I have a few todos scattered about referencing the need for providing context about the user making the request. Given the current state of these abstractions, your solution seems reasonable.

**Comment by tjwebb** [comment]
Conceivably one could use the Waterline adapter specification to shim in other datastores:

- https://github.com/balderdashy/waterline#adapters-concept
- https://github.com/balderdashy/sails-docs/blob/master/contributing/adapter-specification.md

**Comment by tjwebb** [comment]
> Issues are good for specific problems, IRC is probably better for general discussions

FWIW, if you guys were talking about this on IRC, google would not have led me here.
